consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundState.java

```diff
@@ -0,0 +1,125 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.ibft.statemachine;
+
+import tech.pegasys.pantheon.consensus.ibft.ConsensusRoundIdentifier;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.CommitPayload;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.PreparePayload;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.PreparedCertificate;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.ProposalPayload;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.SignedData;
+import tech.pegasys.pantheon.consensus.ibft.validation.MessageValidator;
+import tech.pegasys.pantheon.crypto.SECP256K1.Signature;
+import tech.pegasys.pantheon.ethereum.core.Block;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Sets;
+
+// Data items used to define how a round will operate
+public class RoundState {
+
+  private final ConsensusRoundIdentifier roundIdentifier;
+  private final MessageValidator validator;
+  private final long quorumSize;
+
+  private Optional<SignedData<ProposalPayload>> proposalMessage = Optional.empty();
+
+  // Must track the actual Prepare message, not just the sender, as these may need to be reused
+  // to send out in a PrepareCertificate.
+  private final Set<SignedData<PreparePayload>> preparePayloads = Sets.newHashSet();
+  private final Set<SignedData<CommitPayload>> commitPayloads = Sets.newHashSet();
+
+  private boolean prepared = false;
+  private boolean committed = false;
+
+  public RoundState(
+      final ConsensusRoundIdentifier roundIdentifier,
+      final int quorumSize,
+      final MessageValidator validator) {
+    this.roundIdentifier = roundIdentifier;
+    this.quorumSize = quorumSize;
+    this.validator = validator;
+  }
+
+  public ConsensusRoundIdentifier getRoundIdentifier() {
+    return roundIdentifier;
+  }
+
+  public boolean setProposedBlock(final SignedData<ProposalPayload> msg) {
+    if (validator.addSignedProposalPayload(msg)) {
+      proposalMessage = Optional.of(msg);
+    } else {
+      return false;
+    }
+    preparePayloads.removeIf(p -> !validator.validatePrepareMessage(p));
+    commitPayloads.removeIf(p -> !validator.validateCommmitMessage(p));
+    updateState();
+    return true;
+  }
+
+  public void addPreparedPeer(final SignedData<PreparePayload> prepareMsg) {
+    if (!proposalMessage.isPresent()) {
+      preparePayloads.add(prepareMsg);
+    } else if (validator.validatePrepareMessage(prepareMsg)) {
+      preparePayloads.add(prepareMsg);
+    }
+    updateState();
+  }
+
+  public void addCommitSeal(final SignedData<CommitPayload> commitPayload) {
+    if (!proposalMessage.isPresent()) {
```

Comment:
simplify to (!proposalMessage.isPresent() || validator.validateCommmitMessage(commitPayload))

Reply:
Less lines of code - but potentially harder read - where do other reviewers fall?